x/oracle/abci.go

```diff
@@ -33,32 +33,46 @@ func EndBlocker(ctx sdk.Context, k keeper.Keeper) {
 
 		switch task := task.(type) {
 		case *types.Task:
-			ctx.EventManager().EmitEvent(
-				sdk.NewEvent(
-					"aggregate_task",
-					sdk.NewAttribute("contract", task.Contract),
-					sdk.NewAttribute("function", task.Function),
-					sdk.NewAttribute("begin_block_height", strconv.FormatInt(task.BeginBlock, 10)),
-					sdk.NewAttribute("bounty", task.Bounty.String()),
-					sdk.NewAttribute("description", task.Description),
-					sdk.NewAttribute("expiration", task.Expiration.String()),
-					sdk.NewAttribute("creator", task.Creator),
-					sdk.NewAttribute("responses", task.Responses.String()),
-					sdk.NewAttribute("result", task.Result.String()),
-					sdk.NewAttribute("end_block_height", strconv.FormatInt(task.ExpireHeight, 10)),
-					sdk.NewAttribute("status", task.Status.String()),
-				),
-			)
+			EmitEventsForTask(ctx, task)
 		case *types.TxTask:
-			//implement me
-			ctx.EventManager().EmitEvent(
-				sdk.NewEvent(
-					"aggregate_task",
-					//shall be base64 encoded to be aligned with proto json encoding
-					sdk.NewAttribute("tx_hash", base64.StdEncoding.EncodeToString(task.TxHash)),
-				),
-			)
+			EmitEventsForTxTask(ctx, task)
 		}
 	}
 	k.DeleteClosingTaskIDs(ctx)
+	k.DeleteExpiredTasks(ctx)
+	k.SetLastBlockTime(ctx, ctx.BlockTime())
+}
+
+func EmitEventsForTask(ctx sdk.Context, task *types.Task) {
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			"aggregate_task",
+			sdk.NewAttribute("contract", task.Contract),
+			sdk.NewAttribute("function", task.Function),
+			sdk.NewAttribute("begin_block_height", strconv.FormatInt(task.BeginBlock, 10)),
+			sdk.NewAttribute("bounty", task.Bounty.String()),
+			sdk.NewAttribute("description", task.Description),
+			sdk.NewAttribute("expiration", task.Expiration.String()),
+			sdk.NewAttribute("creator", task.Creator),
+			sdk.NewAttribute("responses", task.Responses.String()),
+			sdk.NewAttribute("result", task.Result.String()),
+			sdk.NewAttribute("end_block_height", strconv.FormatInt(task.ExpireHeight, 10)),
+			sdk.NewAttribute("status", task.Status.String()),
+		),
+	)
+}
+
+func EmitEventsForTxTask(ctx sdk.Context, task *types.TxTask) {
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			"aggregate_txtask",
```

Comment:
`aggregate_tx_task`?
A suggestion but not essential, can be put in events.go

Reply:
it sounds good to me. fixed.